### PR TITLE
Merging Identity v1.2

### PR DESCRIPTION
**Pull request by mikiquantum**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Issue Link: <Github link>

### CODE REVIEW

File: testingutils/utils.go [hunk 1]
```diff
@@ -41,3 +44,9 @@ func (m *MockQueue) EnqueueJobWithMaxTries(taskTypeName string, params map[strin
 	res, _ := args.Get(0).(queue.TaskResult)
 	return res, args.Error(1)
 }
+
+// RandomDID return a random did for testing.
+func RandomDID() identity.DID {
```
Comment (manuelpolzhofer) [inline testingutils/utils.go]: duplicate method use in testingidentity `GenerateRandomDID`

File: identity/ideth/bootstrapper.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/centrifuge/go-centrifuge/ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/savaki/jq"
+)
+
+// Bootstrapper implements bootstrap.Bootstrapper.
+type Bootstrapper struct{}
+
+// BootstrappedDIDFactory stores the id of the factory
+const BootstrappedDIDFactory string = "BootstrappedDIDFactory"
```
Comment (mikiquantum) [inline identity/ideth/bootstrapper.go]: Remove these constants

File: identity/ideth/bootstrapper.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+func bindFactory(factoryAddress common.Address, client ethereum.Client) (*FactoryContract, error) {
+	return NewFactoryContract(factoryAddress, client.GetEthClient())
+}
+
+func getFactoryAddress(cfg config.Configuration) common.Address {
+	return cfg.GetContractAddress(config.IdentityFactory)
+}
+
+// TODO: store identity bytecode in config and remove func
+func getIdentityByteCode() string {
```
Comment (mikiquantum) [inline identity/ideth/bootstrapper.go]: We should remove this, and read from node config

File: identity/ideth/bootstrapper.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		panic(err)
+	}
+	byteCodeHex := getOpField(optByte, dat)
+	return byteCodeHex
+
+}
+
+// TODO: store identity bytecode in config and remove func
+func getOpField(addrOp jq.Op, dat []byte) string {
```
Comment (mikiquantum) [inline identity/ideth/bootstrapper.go]: This too.

File: identity/ideth/service.go [hunk 5]
```diff
@@ -174,13 +118,13 @@ func (i service) getDID(ctx context.Context) (did DID, err error) {
 	if err != nil {
 		return did, err
 	}
-	did = NewDID(common.BytesToAddress(addressByte))
+	did = id.NewDID(common.BytesToAddress(addressByte))
```
Comment (mikiquantum) [inline identity/ideth/service.go]:
```suggestion
	DID = id.NewDID(common.BytesToAddress(addressByte))
```

File: identity/ideth/service.go [hunk 6]
```diff
@@ -174,13 +118,13 @@ func (i service) getDID(ctx context.Context) (did DID, err error) {
...
 	return did, nil
 
 }
 
 // AddKey adds a key to identity contract
-func (i service) AddKey(ctx context.Context, key Key) error {
+func (i service) AddKey(ctx context.Context, key id.KeyDID) error {
 	did, err := i.getDID(ctx)
```
Comment (mikiquantum) [inline identity/ideth/service.go]:
```suggestion
	DID, err := i.getDID(ctx)
```

File: identity/ideth/service.go [hunk 7]
```diff
@@ -452,77 +396,92 @@ func (i service) GetClientsP2PURLs(dids []*DID) ([]string, error) {
...
+	err := i.ValidateKey(context.Background(), centID, signature.PublicKey, id.KeyPurposeSigning)
 	if err != nil {
 		return err
 	}
+
+	return crypto.VerifySignature(signature.PublicKey, message, signature.Signature)
+}
+
+// ValidateCentrifugeIDBytes validates a centrifuge ID given as bytes
+func ValidateCentrifugeIDBytes(givenCentID []byte, centrifugeID id.DID) error {
```
Comment (mikiquantum) [inline identity/ideth/service.go]:
```suggestion
func ValidateCentrifugeIDBytes(givenDID []byte, DID id.DID) error {
```

File: p2p/client_integration_test.go [hunk 8]
```diff
@@ -66,32 +86,40 @@ func TestClient_GetSignaturesForDocumentValidationCheck(t *testing.T) {
...
 		assert.Equal(t, "[1]document is invalid: [mismatched document roots]", err.Error())
 	}
 }
 
-func createLocalCollaborator(t *testing.T, corruptID bool) (*configstore.Account, identity.Identity, error) {
-	tcID := identity.RandomCentID()
+func createLocalCollaborator(t *testing.T, corruptID bool) (*configstore.Account, identity.DID, error) {
+	didAddr, err := idFactory.CalculateIdentityAddress(context.Background())
+	assert.NoError(t, err)
+	did := identity.NewDID(*didAddr)
```
Comment (mikiquantum) [inline p2p/client_integration_test.go]:
```suggestion
	DID := identity.NewDID(*didAddr)
```

File: p2p/receiver/handler.go [hunk 9]
```diff
@@ -47,12 +47,12 @@ func (srv *Handler) HandleInterceptor(ctx context.Context, peer peer.ID, protoc
 		return convertToErrorEnvelop(err)
 	}
 
-	cid, err := p2pcommon.ExtractCID(protoc)
+	did, err := p2pcommon.ExtractDID(protoc)
```
Comment (mikiquantum) [inline p2p/receiver/handler.go]:
```suggestion
	DID, err := p2pcommon.ExtractDID(protoc)
```

**Comment by mikiquantum** [comment]
closing as I will open a new PR from the identity branch

**Pull request closed without merging**